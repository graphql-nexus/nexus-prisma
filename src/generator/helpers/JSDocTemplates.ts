import { DMMF } from '@prisma/client/runtime'
import endent from 'endent'
import { repeat } from 'lodash'

export function jsDocForModel(model: DMMF.Model): string {
  return model.documentation
    ? endent`
        /**
         * ${model.documentation}
         */
      `
    : missingModelDocumentation(model)
}

export function jsDocForField({ field, model }: { field: DMMF.Field; model: DMMF.Model }): string {
  return field.documentation
    ? endent`
        /**
         * ${field.documentation}
         */
      `
    : missingFieldDocumentation({ field, model })
}

export function missingModelDocumentation({ name }: DMMF.Model): string {
  return endent`
    /**
     * ### ðŸ“” Missing Model Documentation for \`${name}\`
     * 
     * Get JSDoc documentation for this model automatically by documenting
     * it in your Prisma Schema âœ¨!
     * 
     * To document a model add a tripple slash comment above it.
     * 
     * Learn more about Prisma Schema comments [here](https://www.prisma.io/docs/concepts/components/prisma-schema#comments).
     * 
     * ${jsDocDiviion('Example 1')}
     * 
     * \`\`\`prisma
     * /// Lorem ipsum dolor sit amet...
     * model ${name} {
     *   id  String  @id
     * }
     * \`\`\`
     * 
     * ${jsDocDiviion('Example 2')}
     * 
     * \`\`\`prisma
     * /// A user record. A user maps 1:1 with a person. Not
     * /// to be confused with an \`Account\` which on person
     * /// may have multiple of, all linked to a single \`User\`.
     * model User {
     *   id  String  @id
     * }
     * \`\`\`
     */
  `
}

export function missingFieldDocumentation({
  field,
  model,
}: {
  field: DMMF.Field
  model: DMMF.Model
}): string {
  return endent`
    /**
     * ### ðŸ“” Missing Field Documentation for \`${field.name}\`
     * 
     * Get JSDoc documentation for this field automatically by documenting
     * it in your Prisma Schema âœ¨!
     * 
     * To document a field add a tripple slash comment above it.
     * 
     * Learn more about Prisma Schema comments [here](https://www.prisma.io/docs/concepts/components/prisma-schema#comments).
     * 
     * ${jsDocDiviion('Example 1')}
     * 
     * \`\`\`prisma
     * model ${model.name} {
     *   /// Lorem ipsum dolor sit amet.
     *   ${field.name}  ${field.type}${field.isRequired ? '' : '?'}
     * }
     * \`\`\`
     * 
     * ${jsDocDiviion('Example 2')}
     * 
     * \`\`\`prisma
     * model User {
     *   /// Identifier for a user. Will never change once set.
     *   /// More stable than the \`handle\` field  which user
     *   /// _can_ change, but is not human-friendly. Prefer this
     *   /// for machine consumers but prefer \`handle\` for human
     *   /// visible things like URL slugs.
     *   id  String  @id
     * }
     * \`\`\`
     */
  `
}

/**
 * Render a centered title with bars on either side filling left over space.
 *
 * Sized according to standard JSDoc window size in VSCode.
 */
function jsDocDiviion(title: string): string {
  const titlePadding = ' '
  const barLength = Math.floor(MAX_JS_DOC_LENGTH / 2 - title.length - titlePadding.length * 2)
  const bar = repeat('â€“', barLength)
  return endent`
    #### ${bar}${titlePadding}${title}${titlePadding}${bar}
  `
}

const MAX_JS_DOC_LENGTH = 72
