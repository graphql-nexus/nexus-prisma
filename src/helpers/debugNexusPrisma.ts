import debug from 'debug'

const debugName = 'nexus-prisma'

export const d = debug(debugName)

d.log = console.log.bind(console)

export function debugNexusPrisma(...names: string[]) {
  const d2 = debug([debugName].concat(names).join(':'))
  d2.log = console.log.bind(console)
  return d2
}

export function packageRun() {}
