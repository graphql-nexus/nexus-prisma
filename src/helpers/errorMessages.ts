export const GITHUB_NEW_ISSUE_LINK = `https://github.com/prisma/nexus-prisma/issues/new/choose`

export const GITHUB_NEW_DISCUSSION_LINK = `https://github.com/prisma/nexus-prisma/discussions/new`
