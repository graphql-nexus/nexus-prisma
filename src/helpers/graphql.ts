export type GraphQLScalarType = 'ID' | 'String' | 'Int' | 'Float' | 'Boolean'

export const graphQLScalarTypes = {
  ID: 'ID',
  String: 'String',
  Float: 'Float',
  Boolean: 'Boolean',
  Int: 'Int',
} as const
