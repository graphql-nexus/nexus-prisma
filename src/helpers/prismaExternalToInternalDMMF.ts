// @ts-expect-error Private api
import * as PrismaClientGenerator from '@prisma/client/generator-build'
import { DMMF, DMMF as ExternalDMMF } from '@prisma/generator-helper'

export const externalToInternalDmmf = PrismaClientGenerator.externalToInternalDmmf as (
  document: ExternalDMMF.Document
) => DMMF.Document
