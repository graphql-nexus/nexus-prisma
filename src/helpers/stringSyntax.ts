import endent from 'endent'

export function quote(x: string): string {
  return `'${x}'`
}

export function apply(functionName: string, parameters: string | string[]): string {
  return endent`
    ${functionName}(${arrayify(parameters).join(', ')})
  `
}

function arrayify<T>(x: T): T extends any[] ? T : T[] {
  if (Array.isArray(x)) {
    return x as any
  }
  return [x] as any
}
