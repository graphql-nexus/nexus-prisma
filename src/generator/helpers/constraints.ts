import { DMMF } from '@prisma/client/runtime'
import { RecordUnknown } from '../../helpers/utils'

/**
 * Find the unique identifiers necessary to indentify a field
 *
 * Unique fields for a model can be one of (in this order):
 * 1. One (and only one) field with an @id annotation
 * 2. Multiple fields with @@id clause
 * 3. One (and only one) field with a @unique annotation (if there are multiple, use the first one)
 * 4. Multiple fields with a @@unique clause
 */
export function resolveUniqueIdentifiers(model: DMMF.Model): string[] {
  // Try finding 1.
  const singleIdField = model.fields.find((f) => f.isId)

  if (singleIdField) {
    return [singleIdField.name]
  }

  // Try finding 2.
  if (model.idFields && model.idFields.length > 0) {
    return model.idFields
  }

  // Try finding 3.
  const singleUniqueField = model.fields.find((f) => f.isUnique)

  if (singleUniqueField) {
    return [singleUniqueField.name]
  }

  // Try finding 4.
  if (model.uniqueFields && model.uniqueFields.length > 0) {
    return model.uniqueFields[0] as string[] // I don't know why typescript want a cast here
  }

  throw new Error(`Unable to resolve a unique identifier for the Prisma model: ${model.name}`)
}

export function findMissingUniqueIdentifiers(
  data: RecordUnknown,
  uniqueIdentifiers: string[]
): string[] | null {
  const missingIdentifiers: string[] = []

  for (const identifier of uniqueIdentifiers) {
    if (!data[identifier]) {
      missingIdentifiers.push(identifier)
    }
  }

  if (missingIdentifiers.length > 0) {
    return missingIdentifiers
  }

  return null
}

export function buildWhereUniqueInput(data: RecordUnknown, uniqueIdentifiers: string[]): RecordUnknown {
  if (uniqueIdentifiers.length === 1) {
    return pickFromRecord(data, uniqueIdentifiers)
  }

  const compoundName = uniqueIdentifiers.join('_')

  return {
    [compoundName]: pickFromRecord(data, uniqueIdentifiers),
  }
}

function pickFromRecord(record: RecordUnknown, keys: string[]) {
  const output: Record<string, unknown> = {}

  for (const identifier of keys) {
    if (record[identifier]) {
      output[identifier] = record[identifier]
    }
  }

  return output
}
