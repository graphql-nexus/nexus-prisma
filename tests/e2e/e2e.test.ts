import debug from 'debug'
import dedent from 'dindist'
import * as Execa from 'execa'
import { gql } from 'graphql-request'
import stripAnsi from 'strip-ansi'
import { inspect } from 'util'
import { assertBuildPresent } from '../__helpers__/helpers'
import { createPrismaSchema } from '../__helpers__/testers'
import { setupTestProject, TestProject } from '../__helpers__/testProject'
import * as GQLScalars from 'graphql-scalars'

const d = debug('e2e')

interface FileSpec {
  filePath: string
  content: string
}

interface ProjectResult {
  runFirstBuild: Execa.ExecaSyncReturnValue<string>
  runReflectPrisma: Execa.ExecaSyncReturnValue<string>
  runReflectNexus: Execa.ExecaSyncReturnValue<string>
  runSecondBuild: Execa.ExecaSyncReturnValue<string>
  fileGraphqlSchema?: string
  fileTypegen?: string
}

const TYPEGEN_FILE_NAME = `typegen.ts`
const TYPEGEN_FILE_PATH = `src/${TYPEGEN_FILE_NAME}`

const GRAPHQL_SCHEMA_FILE_PATH = `schema.graphql`

const SERVER_READY_MESSAGE = `GraphQL API ready at http://localhost:4000/graphql`

function runTestProjectBuild(testProject: TestProject): ProjectResult {
  const commandConfig: Execa.SyncOptions = {
    reject: false,
    cwd: testProject.fs.cwd(),
  }
  const runFirstBuild = Execa.commandSync(`npm run build`, commandConfig)
  const runReflectPrisma = Execa.commandSync(`npm run reflect:prisma`, commandConfig)
  const runReflectNexus = Execa.commandSync(`npm run reflect:nexus`, commandConfig)
  const runSecondBuild = Execa.commandSync(`npm run build`, commandConfig)
  const fileGraphqlSchema = testProject.fs.read(GRAPHQL_SCHEMA_FILE_PATH)
  const fileTypegen = testProject.fs.read(TYPEGEN_FILE_PATH)

  return {
    runFirstBuild,
    runReflectPrisma,
    runReflectNexus,
    runSecondBuild,
    fileGraphqlSchema,
    fileTypegen,
  }
}

function setupTestNexusPrismaProject(): TestProject {
  const testProject = setupTestProject({
    tsconfigJson: {},
    packageJson: {
      license: 'MIT',
      scripts: {
        reflect: 'yarn -s reflect:prisma && yarn -s reflect:nexus',
        'reflect:prisma': "cross-env DEBUG='*' prisma generate",
        // peer dependency check will fail since we're using yalc, e.g.:
        // " ... nexus-prisma@0.0.0-dripip+c2653557 does not officially support @prisma/client@2.22.1 ... "
        'reflect:nexus': 'cross-env REFLECT=true ts-node --transpile-only src/schema',
        build: 'tsc',
        start: 'node build/server',
        'dev:server': 'yarn ts-node-dev --transpile-only server',
        'db:migrate': 'prisma db push --force-reset && ts-node prisma/seed',
      },
      dependencies: {
        dotenv: '^9.0.0',
        'apollo-server': '^2.24.0',
        'cross-env': '^7.0.1',
        '@prisma/client': '^2.18.0',
        '@types/node': '^14.14.32',
        graphql: '^15.5.0',
        nexus: '^1.0.0',
        prisma: '^2.18.0',
        'ts-node': '^9.1.1',
        'ts-node-dev': '^1.1.6',
        typescript: '^4.2.3',
      },
    },
  })

  if (testProject.info.isReusing) {
    d(`starting project setup cleanup for reuse`)
    testProject.fs.remove(TYPEGEN_FILE_PATH)
    testProject.fs.remove('node_modules/nexus-prisma')
    testProject.runOrThrow(`yalc add nexus-prisma`)
    d(`done project setup cleanup for reuse`)
  } else {
    d(`starting project setup`)
    Execa.commandSync(`yalc publish --no-scripts`, { stdio: 'inherit' })
    testProject.runOrThrow(`yalc add nexus-prisma`, { stdio: 'inherit' })
    testProject.runOrThrow(`npm install`, { stdio: 'inherit' })
    d(`done project setup`)
  }

  return testProject
}

let testProject: TestProject

beforeAll(() => {
  assertBuildPresent()
  testProject = setupTestNexusPrismaProject()
})

it('When bundled custom scalars are used the project type checks and generates expected GraphQL schema', async () => {
  const files: FileSpec[] = [
    {
      filePath: `prisma/schema.prisma`,
      content: createPrismaSchema({
        content: dedent`
          model Foo {
            id                String   @id
            someJsonField     Json
            someDateTimeField DateTime
            someDecimalField  Decimal
            someBytesField    Bytes
            someBigIntField   BigInt
            someEnumA         SomeEnumA
            bar               Bar?
          }

          model Bar {
            id    String  @id
            foo   Foo?    @relation(fields: [fooId], references: [id])
            fooId String?
          }

          enum SomeEnumA {
            alpha
            bravo
            charlie
          }
        `,
      }),
    },
    {
      filePath: `prisma/seed.ts`,
      content: dedent/*ts*/ `
        import { PrismaClient, Prisma } from '@prisma/client'

        main()

        async function main() {
          const prisma = new PrismaClient()

          await prisma.$executeRaw('TRUNCATE "Foo", "Bar"')
          await prisma.foo.create({
            data: {
              id: 'foo1',
              someDateTimeField: new Date("2021-05-10T20:42:46.609Z"),
              someDecimalField: new Prisma.Decimal(24.454545),
              someBytesField: Buffer.from([]),
              someJsonField: {},
              someBigIntField: BigInt(9007199254740991),
              someEnumA: 'alpha',
              bar: {
                create: {
                  id: 'bar1',
                },
              },
            },
          })

          await prisma.$disconnect()
        }
      `,
    },
    {
      filePath: `src/schema.ts`,
      content: dedent/*ts*/ `
        require('dotenv').config()

        import { makeSchema, objectType, enumType, queryType } from 'nexus'
        import { Bar, Foo, SomeEnumA, $settings } from 'nexus-prisma'
        import * as customScalars from 'nexus-prisma/scalars'
        import * as Path from 'path'

        // Show that we can import and call settings as a NOOP
        $settings({})
        
        const types = [
          customScalars,
          enumType(SomeEnumA),
          queryType({
            definition(t) {
              t.list.field({
                name: 'bars',
                type: 'Bar',
                resolve(_, __, ctx) {
                  return ctx.prisma.bar.findMany()
                },
              })
            },
          }),
          objectType({
            name: Bar.$name,
            definition(t) {
              t.field(Bar.foo)
            },
          }),
          objectType({
            name: Foo.$name,
            definition(t) {
              t.field({
                name: 'someEnumA',
                type: 'SomeEnumA',
              })
              t.json('JsonManually')
              t.dateTime('DateTimeManually')
              t.bytes('BytesManually')
              t.decimal('DecimalManually')
              t.bigInt('BigIntManually')
              t.field(Foo.someBigIntField)
              t.field(Foo.someJsonField)
              t.field(Foo.someDateTimeField)
              t.field(Foo.someBytesField)
              t.field(Foo.someDecimalField)
            },
          }),
        ]
        
        const schema = makeSchema({
          types,
          shouldGenerateArtifacts: true,
          shouldExitAfterGenerateArtifacts: Boolean(process.env.REFLECT),
          outputs: {
            schema: true,
            typegen: Path.join(__dirname, '${TYPEGEN_FILE_NAME}'),
          },
          sourceTypes: {
            modules: [{ module: '.prisma/client', alias: 'PrismaClient' }],
          },
        })

        export default schema
      `,
    },
    {
      filePath: `src/context.ts`,
      content: dedent/*ts*/ `
        import { PrismaClient } from '@prisma/client'

        const prisma = new PrismaClient()
        
        type Context = {
          prisma: PrismaClient
        }
        
        function context() {
          return {
            prisma,
          }
        }

        export default context
        export { Context }
      `,
    },
    {
      filePath: `src/server.ts`,
      content: dedent/*ts*/ `
        require('dotenv').config()

        import { ApolloServer } from 'apollo-server'
        import context from './context'
        import schema from './schema'

        const apollo = new ApolloServer({
          schema,
          context,
        })

        apollo.listen(4000, () => {
          console.log('${SERVER_READY_MESSAGE}')
        })
      `,
    },
    {
      filePath: `.env`,
      content: dedent`
        DB_URL="postgres://bcnfshogmxsukp:e31b6ddc8b9d85f8964b6671e4b578c58f0d13e15f637513207d44268eabc950@ec2-54-196-33-23.compute-1.amazonaws.com:5432/d17vadgam0dtao?schema=${
          process.env.E2E_DB_SCHEMA ?? 'local'
        }"
        NO_PEER_DEPENDENCY_CHECK="true"
      `,
    },
  ]

  files.forEach((fileSpec) => testProject.fs.write(fileSpec.filePath, fileSpec.content))

  // todo api server & database & seed that allows for testing that prisma runtime usage works

  // uncomment this to see dir (helpful to go there yourself and manually debug)
  console.log(`e2e test project at: ${testProject.fs.cwd()}`)

  const results = runTestProjectBuild(testProject)

  // uncomment this to see the raw results (helpful for debugging)
  console.log(`e2e output:\n`, inspect(results, { depth: 10, colors: true }))

  /**
   * Sanity checks around buildtime
   */

  expect(results.runFirstBuild.exitCode).toBe(2)

  expect(stripAnsi(results.runFirstBuild.stdout)).toMatch(
    /.*error TS2305: Module '"nexus-prisma"' has no exported member '\$settings'.*/
  )
  expect(stripAnsi(results.runFirstBuild.stdout)).toMatch(
    /.*error TS2305: Module '"nexus-prisma"' has no exported member 'Bar'.*/
  )
  expect(stripAnsi(results.runFirstBuild.stdout)).toMatch(
    /.*error TS2305: Module '"nexus-prisma"' has no exported member 'Foo'.*/
  )
  expect(stripAnsi(results.runFirstBuild.stdout)).toMatch(
    /.*error TS2305: Module '"nexus-prisma"' has no exported member 'SomeEnumA'.*/
  )
  expect(stripAnsi(results.runFirstBuild.stdout)).toMatch(
    /.*error TS2339: Property 'json' does not exist on type 'ObjectDefinitionBlock<any>'.*/
  )
  expect(stripAnsi(results.runFirstBuild.stdout)).toMatch(
    /.*error TS2339: Property 'dateTime' does not exist on type 'ObjectDefinitionBlock<any>'.*/
  )
  expect(stripAnsi(results.runFirstBuild.stdout)).toMatch(
    /.*error TS2339: Property 'bigInt' does not exist on type 'ObjectDefinitionBlock<any>'.*/
  )

  expect(results.runReflectPrisma.exitCode).toBe(0)

  expect(stripAnsi(results.runReflectPrisma.stdout)).toMatch(/.*Generated Nexus Prisma.*/)

  expect(results.runReflectNexus.exitCode).toBe(0)

  expect(stripAnsi(results.runReflectNexus.stdout)).toMatch(/.*Generated Artifacts.*/)

  expect(results.runSecondBuild.exitCode).toBe(0)

  expect(results.fileGraphqlSchema).toMatchSnapshot('graphql schema')

  expect(results.fileTypegen).toMatchSnapshot('nexus typegen')

  /**
   * Sanity check the Prisma Client import ID
   */

  expect(testProject.fs.read('node_modules/nexus-prisma/dist/runtime/index.js')).toMatch(
    /.*"prismaClientImportId": "@prisma\/client".*/
  )

  /**
   * Sanity check the runtime
   */

  d(`migrating database`)

  testProject.runOrThrow(`npm run db:migrate`)

  d(`starting server`)

  const serverProcess = testProject.runAsync(`node build/server`, { reject: false })
  serverProcess.stdout!.pipe(process.stdout)

  await new Promise((res) =>
    serverProcess.stdout!.on('data', (data: Buffer) => {
      if (data.toString().match(SERVER_READY_MESSAGE)) res(undefined)
    })
  )

  d(`starting client queries`)

  const data = await testProject.client.request(gql`
    query {
      bars {
        foo {
          JsonManually
          DateTimeManually
          BytesManually
          DecimalManually
          BigIntManually
          someEnumA
          someJsonField
          someDateTimeField
          someBytesField
          someDecimalField
          someBigIntField
        }
      }
    }
  `)

  d(`stopping server`)

  serverProcess.cancel()
  // On Windows the serverProcess never completes the promise so we do an ugly timeout here
  // and rely on jest --forceExit to terminate the process
  await Promise.race([serverProcess, new Promise((res) => setTimeout(res, 2000))])

  d(`stopped server`)

  expect(data).toMatchSnapshot('client request 1')

  const [{ foo }] = data.bars

  expect(foo.JsonManually).toBeNull()
  expect(foo.DateTimeManually).toBeNull()
  expect(foo.BytesManually).toBeNull()
  expect(foo.BigIntManually).toBeNull()
  expect(typeof foo.someEnumA).toEqual('string')
  expect(typeof foo.someJsonField).toEqual('object')
  expect(typeof foo.someDateTimeField).toEqual('string')
  expect(typeof foo.someBytesField).toEqual('object')
  expect(typeof GQLScalars.BigIntResolver.parseValue(foo.someBigIntField)).toEqual('bigint')
}, 60_000)
