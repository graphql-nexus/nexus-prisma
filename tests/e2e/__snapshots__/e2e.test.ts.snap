// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When bundled custom scalars are used the project type checks and generates expected GraphQL schema: client request 1 1`] = `
Object {
  "bars": Array [
    Object {
      "foo": Object {
        "BigIntManually": null,
        "BytesManually": null,
        "DateTimeManually": null,
        "DecimalManually": null,
        "JsonManually": null,
        "someBigIntField": "9007199254740991",
        "someBytesField": Object {
          "data": Array [],
          "type": "Buffer",
        },
        "someDateTimeField": "2021-05-10T20:42:46.609Z",
        "someDecimalField": "24.454545",
        "someEnumA": "alpha",
        "someJsonField": Object {},
      },
    },
  ],
}
`;

exports[`When bundled custom scalars are used the project type checks and generates expected GraphQL schema: graphql schema 1`] = `
"### This file was generated by Nexus Schema
### Do not make changes to this file directly


type Bar {
  foo: Foo
}

\\"\\"\\"
The \`BigInt\` scalar type represents non-fractional signed whole numeric values.
@see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt
\\"\\"\\"
scalar BigInt

\\"\\"\\"The \`Byte\` scalar type represents byte value as a Buffer\\"\\"\\"
scalar Bytes

\\"\\"\\"
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
\\"\\"\\"
scalar DateTime

\\"\\"\\"An arbitrary-precision Decimal type\\"\\"\\"
scalar Decimal

type Foo {
  BigIntManually: BigInt
  BytesManually: Bytes
  DateTimeManually: DateTime
  DecimalManually: Decimal
  JsonManually: Json
  someBigIntField: BigInt!
  someBytesField: Bytes!
  someDateTimeField: DateTime!
  someDecimalField: Decimal!
  someEnumA: SomeEnumA
  someJsonField: Json!
}

\\"\\"\\"
The \`JSONObject\` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
\\"\\"\\"
scalar Json @specifiedBy(url: \\"http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf\\")

type Query {
  bars: [Bar]
}

enum SomeEnumA {
  alpha
  bravo
  charlie
}
"
`;

exports[`When bundled custom scalars are used the project type checks and generates expected GraphQL schema: nexus typegen 1`] = `
"/**
 * This file was generated by Nexus Schema
 * Do not make changes to this file directly
 */

import type * as PrismaClient from \\".prisma/client\\"
import type { core } from \\"nexus\\"
declare global {
  interface NexusGenCustomInputMethods<TypeName extends string> {
    /**
     * The \`BigInt\` scalar type represents non-fractional signed whole numeric values.
     * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt
     */
    bigInt<FieldName extends string>(fieldName: FieldName, opts?: core.CommonInputFieldConfig<TypeName, FieldName>): void // \\"BigInt\\";
    /**
     * The \`Byte\` scalar type represents byte value as a Buffer
     */
    bytes<FieldName extends string>(fieldName: FieldName, opts?: core.CommonInputFieldConfig<TypeName, FieldName>): void // \\"Bytes\\";
    /**
     * A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
     */
    dateTime<FieldName extends string>(fieldName: FieldName, opts?: core.CommonInputFieldConfig<TypeName, FieldName>): void // \\"DateTime\\";
    /**
     * An arbitrary-precision Decimal type
     */
    decimal<FieldName extends string>(fieldName: FieldName, opts?: core.CommonInputFieldConfig<TypeName, FieldName>): void // \\"Decimal\\";
    /**
     * The \`JSONObject\` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
     */
    json<FieldName extends string>(fieldName: FieldName, opts?: core.CommonInputFieldConfig<TypeName, FieldName>): void // \\"Json\\";
  }
}
declare global {
  interface NexusGenCustomOutputMethods<TypeName extends string> {
    /**
     * The \`BigInt\` scalar type represents non-fractional signed whole numeric values.
     * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt
     */
    bigInt<FieldName extends string>(fieldName: FieldName, ...opts: core.ScalarOutSpread<TypeName, FieldName>): void // \\"BigInt\\";
    /**
     * The \`Byte\` scalar type represents byte value as a Buffer
     */
    bytes<FieldName extends string>(fieldName: FieldName, ...opts: core.ScalarOutSpread<TypeName, FieldName>): void // \\"Bytes\\";
    /**
     * A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the \`date-time\` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
     */
    dateTime<FieldName extends string>(fieldName: FieldName, ...opts: core.ScalarOutSpread<TypeName, FieldName>): void // \\"DateTime\\";
    /**
     * An arbitrary-precision Decimal type
     */
    decimal<FieldName extends string>(fieldName: FieldName, ...opts: core.ScalarOutSpread<TypeName, FieldName>): void // \\"Decimal\\";
    /**
     * The \`JSONObject\` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
     */
    json<FieldName extends string>(fieldName: FieldName, ...opts: core.ScalarOutSpread<TypeName, FieldName>): void // \\"Json\\";
  }
}


declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
  SomeEnumA: PrismaClient.SomeEnumA
}

export interface NexusGenScalars {
  String: string
  Int: number
  Float: number
  Boolean: boolean
  ID: string
  BigInt: any
  Bytes: any
  DateTime: any
  Decimal: any
  Json: any
}

export interface NexusGenObjects {
  Bar: PrismaClient.Bar;
  Foo: PrismaClient.Foo;
  Query: {};
}

export interface NexusGenInterfaces {
}

export interface NexusGenUnions {
}

export type NexusGenRootTypes = NexusGenObjects

export type NexusGenAllTypes = NexusGenRootTypes & NexusGenScalars & NexusGenEnums

export interface NexusGenFieldTypes {
  Bar: { // field return type
    foo: NexusGenRootTypes['Foo'] | null; // Foo
  }
  Foo: { // field return type
    BigIntManually: NexusGenScalars['BigInt'] | null; // BigInt
    BytesManually: NexusGenScalars['Bytes'] | null; // Bytes
    DateTimeManually: NexusGenScalars['DateTime'] | null; // DateTime
    DecimalManually: NexusGenScalars['Decimal'] | null; // Decimal
    JsonManually: NexusGenScalars['Json'] | null; // Json
    someBigIntField: NexusGenScalars['BigInt']; // BigInt!
    someBytesField: NexusGenScalars['Bytes']; // Bytes!
    someDateTimeField: NexusGenScalars['DateTime']; // DateTime!
    someDecimalField: NexusGenScalars['Decimal']; // Decimal!
    someEnumA: NexusGenEnums['SomeEnumA'] | null; // SomeEnumA
    someJsonField: NexusGenScalars['Json']; // Json!
  }
  Query: { // field return type
    bars: Array<NexusGenRootTypes['Bar'] | null> | null; // [Bar]
  }
}

export interface NexusGenFieldTypeNames {
  Bar: { // field return type name
    foo: 'Foo'
  }
  Foo: { // field return type name
    BigIntManually: 'BigInt'
    BytesManually: 'Bytes'
    DateTimeManually: 'DateTime'
    DecimalManually: 'Decimal'
    JsonManually: 'Json'
    someBigIntField: 'BigInt'
    someBytesField: 'Bytes'
    someDateTimeField: 'DateTime'
    someDecimalField: 'Decimal'
    someEnumA: 'SomeEnumA'
    someJsonField: 'Json'
  }
  Query: { // field return type name
    bars: 'Bar'
  }
}

export interface NexusGenArgTypes {
}

export interface NexusGenAbstractTypeMembers {
}

export interface NexusGenTypeInterfaces {
}

export type NexusGenObjectNames = keyof NexusGenObjects;

export type NexusGenInputNames = never;

export type NexusGenEnumNames = keyof NexusGenEnums;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = keyof NexusGenScalars;

export type NexusGenUnionNames = never;

export type NexusGenObjectsUsingAbstractStrategyIsTypeOf = never;

export type NexusGenAbstractsUsingStrategyResolveType = never;

export type NexusGenFeaturesConfig = {
  abstractTypeStrategies: {
    isTypeOf: false
    resolveType: true
    __typename: false
  }
}

export interface NexusGenTypes {
  context: any;
  inputTypes: NexusGenInputs;
  rootTypes: NexusGenRootTypes;
  inputTypeShapes: NexusGenInputs & NexusGenEnums & NexusGenScalars;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  fieldTypeNames: NexusGenFieldTypeNames;
  allTypes: NexusGenAllTypes;
  typeInterfaces: NexusGenTypeInterfaces;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractTypeMembers: NexusGenAbstractTypeMembers;
  objectsUsingAbstractStrategyIsTypeOf: NexusGenObjectsUsingAbstractStrategyIsTypeOf;
  abstractsUsingStrategyResolveType: NexusGenAbstractsUsingStrategyResolveType;
  features: NexusGenFeaturesConfig;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
  }
  interface NexusGenPluginInputTypeConfig<TypeName extends string> {
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
  }
  interface NexusGenPluginInputFieldConfig<TypeName extends string, FieldName extends string> {
  }
  interface NexusGenPluginSchemaConfig {
  }
  interface NexusGenPluginArgConfig {
  }
}"
`;
